### ISSUE TITLE
Message's translation key sent when translation is missing

### ISSUE THREAD

**Issue by abbyad** [medic/medic#4598]
If a translation is missing for a scheduled message's translation key, we are currently sending the key itself as the SMS. Instead we should make sure to not send a message at all.

As part of the [Standard Child Health messaging](https://github.com/medic/medic-projects/issues/4155) we are setting up 60 months worth of monthly messages. If we decide at a later time to send messages only every second month, we would configure the schedule for new registrations accordingly. For existing registrations we would want to be able to blank out the messages that should not go out. We could most easily do that by setting the corresponding translations to blank, or remove them altogether. 

Setting messages to blank seems like a quicker solution than recalculating the entire schedule for previous schedules, and this method can buy us time until the messages are generated on the fly by a rules engine.

Here is an example of a report that has a message pending with the translation key as the message:
```
	"scheduled_tasks": [{
		"due": "2018-06-03T19:27:03.551Z",
		"group": 1,
		"type": "Child_Health",
		"translation_key": "schedule.child_health",
		"message_key": "messages.schedule.child.month_00",
		"recipient": "clinic",
		"state_history": [{
			"state": "scheduled",
			"timestamp": "2018-06-03T19:25:09.804Z"
		},
		{
			"state": "pending",
			"timestamp": "2018-06-03T19:31:00.542Z"
		}],
		"state": "pending",
		"messages": [{
			"uuid": "4a130a0e-386a-4f5d-9d2f-5c50f674aa1b",
			"to": "+14165551230",
			"message": "messages.schedule.child.month_00"
		}]
	},
```

**Comment by abbyad** [comment]
I wanted to reconfirm for myself that we couldn't just have the translation defined but left blank. In the screenshot below a blank translation key shows a blank message while `scheduled`. Once the message goes to `pending` the translation key itself is set into the message.

![image](https://user-images.githubusercontent.com/789512/40929895-10dec9a6-67f4-11e8-9ee7-9c3dd03bda75.png)

**Comment by abbyad** [comment]
Ah, so there isn't an easy way to know that there isn't a translation for the key nor a hardcoded message?

**Comment by garethbowen** [comment]
We know if the translation isn't found. What we don't know is if the configured string was intended to be a translation key or a hardcoded message.

**Comment by abbyad** [comment]
Would a convention here make this better? For instance we could say that for messages all translation keys must not contain a space, whereas all hardcoded messages must contain a space. I could check existing configs, but think that they all already meet that convention because the notation for translation keys is something like `message.form.response_type`

**Comment by abbyad** [comment]
While doing that check I'll also see how many active projects still use hardcoded messages.

**Comment by wtekeu** [comment]
@garethbowen do you still have concerns with @abbyad suggestion to assume that translation keys do not have spaces?

**Comment by garethbowen** [comment]
I have some reservations but I can't think of a better way...

**Comment by abbyad** [comment]
Your other suggestion @garethbowen was to use a special key to say "do not show/send this message" eg `@@@DO_NOT_RENDER@@@`

**Comment by wtekeu** [comment]
@abbyad I have discussed with @dianabarsan about her work on https://github.com/medic/medic-webapp/pull/5052 and I can mark the message as errored and it won't be sent. The challenge is to identify when the translation fails because there are no translations and not because it is a regular message and not a translation key.

**Comment by wtekeu** [comment]
@garethbowen what do you think of the following? 
Currently, it is not possible to edit a non-translatable message given https://github.com/medic/medic-webapp/pull/4872.
So the messages are either hardcoded message that have been added and in that case we have a task that looks like `    {
      "messages": [
        {
          "to": "+256772418764",
          "message": "Test Group 3"
        }
      ],
      "state": "scheduled",
      "group": 3,
      "type": "ANC Reminders LMP",
      "due": "2018-12-10T16:30:00.000Z"
    }` or the task has a `message_key` like `    {
      "due": "2019-01-28T10:00:00.000Z",
      "group": 4,
      "type": "ANC Reminders LMP",
      "translation_key": "schedule.anc_lmp",
      "message_key": "messages.schedule.anc.followup",
      "recipient": "clinic",
      "state_history": [
        {
          "state": "scheduled",
          "timestamp": "2018-10-03T15:35:43.168Z"
        }
      ],
      "state": "scheduled"
    }` therefore it is possible to identify when a translation fails (the task has a field `message_key` and the translation using this translation key fails).

**Comment by garethbowen** [comment]
@wtekeu Yes I think you're right. If the message_key exists we expect it to be translated.

**Comment by abbyad** [comment]
So it seems as though we could go the route of empty translations not being sent, which is great. What would happen to labels that (accidentally) include whitespace, eg `message= ` vs `message= `? 

Should we also have the option of a special key that is explicitly set?

**Comment by wtekeu** [comment]
@abbyad I am not sure I understand your question because the two examples that you gave a perfectly identical (I think).
I just completed the implementation and I assume that if the `message_key` exists and the message returned is exactly equal to the `message_key` then I error the message and it will never be sent.

<img width="753" alt="screen shot 2018-12-06 at 12 24 48 pm" src="https://user-images.githubusercontent.com/1445164/49602282-1800a800-f956-11e8-896d-d6baf9f9d169.png">

**Comment by abbyad** [comment]
They look very similar, but one was intended to have extra whitespace after.

**Comment by wtekeu** [comment]
It looks like we trim so the end results should be the same.

**Comment by newtewt** [comment]
@wtekeu  Hey could you summarize what was changed for this ticket. I see a bit of back and forth just want to ensure testing of the right things here.

**Comment by wtekeu** [comment]
@newtewt if the `message_key` exists and the message returned is exactly equal to the `message_key` then I error the message and it will never be sent.
<img width="753" alt="49602282-1800a800-f956-11e8-896d-d6baf9f9d169" src="https://user-images.githubusercontent.com/1445164/52598531-05a65d00-2e24-11e9-9cba-68eb4bf29fd7.png">

**Comment by ngaruko** [comment]
LGTM
Tested on `gamma.dev` v3.4 by a) setting outgoing language to a language with no available translations and b) deleting some translations...
![image](https://user-images.githubusercontent.com/6979995/53381398-7da08580-39d5-11e9-91c6-f252946ed2d5.png)

Two things however:
a) I think it would be a good idea I think to display a more explicit message to the user like 'Missing translation' instead of 'Message is empty' to help them get it sorted if they missed a translation...what do you think @wtekeu ?

b)There is a bit of an erratic behavior if the outgoing language is set to a language other than English. Hard to put a finger on what is wrong exactly but at first, messages are displayed as expected 
![image](https://user-images.githubusercontent.com/6979995/53382037-b93c4f00-39d7-11e9-978e-5b9c5f405d3c.png)

But then if the page is refreshed/reloaded, it behaves as if no translations exist...(repeatable but hard to guess why this is happening)
![image](https://user-images.githubusercontent.com/6979995/53382228-6dd67080-39d8-11e9-8627-1ea52a3f49fb.png)

**Comment by wtekeu** [comment]
@ngaruko Fixed a) (change the message as you suggested) .. did not fix b) .. according to @dianabarsan :
<img width="874" alt="screen shot 2019-02-27 at 5 41 23 pm" src="https://user-images.githubusercontent.com/1445164/53528362-705bc780-3ab7-11e9-988a-5be7dc476b51.png">

**Comment by ngaruko** [comment]
Thanks @dianabarsan @wtekeu . Will close this and if the language issue is concern, we will raise a different issue.
